### PR TITLE
feat(ngTransclude & $compile): fix treatment of unfilled optional slots

### LINKED ISSUE #13426: How do you tell if an optional multi transclude property has been set?
Basically I can not find a way to check if an optional transclude was supplied, I want to show hide the parent based on the transclude being set or not.

If there is no way is it possible to get this added? It is a useful feature to along side optional transclude.

Thanks

### PR DESCRIPTION
**Pull request by petebacondarwin**
Closes #13426

### CODE REVIEW

**Comment by petebacondarwin** [comment]
@PascalPrecht - hey look at this!

**Comment by PascalPrecht** [comment]
@petebacondarwin awesome! Does this also keep the original contents of the DOM where we transclude into?

E.g.
```
<span ng-transclude="fooSlot">This is default DOM and going to be replaced if `fooSlot` is defined, if not this stays here</span>
```

**Comment by petebacondarwin** [comment]
Yep! That is the second commit in the PR.

**Comment by PascalPrecht** [comment]
Argh. I just've just read the commit messages instead of reading the code haha. Awesome! Thanks for the work!

**Comment by petebacondarwin** [comment]
See https://ci.angularjs.org/job/angular.js-pete/989/artifact/build/docs/api/ng/directive/ngTransclude#multi-slot-transclusion
for a running example

File: src/ng/compile.js [hunk 1]
```diff
@@ -2222,13 +2235,15 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           }
           if (slotName) {
             var slotTranscludeFn = boundTranscludeFn.$$slots[slotName];
-            if (!slotTranscludeFn) {
+            if (slotTranscludeFn) {
+              return slotTranscludeFn(scope, cloneAttachFn, transcludeControllers, futureParentElement, scopeToChild);
+            }
+            if (isUndefined(slotTranscludeFn)) {
```
Comment (Narretz) [inline src/ng/compile.js]: how about just `else`?

**Comment by Narretz** [comment]
`feat($compile): expose transclusion `$$slots` on the `$transclude`
Isn't the commit message a bit contradictory? It's called a feature but `$$slots` indicates a private API.

And a question, is it possible to keep the content with the default behavior (no slots, transclude is true or content)? Since people have requested something similar for the default ngTransclude: https://github.com/angular/angular.js/pull/11839
I think we could extend the 'content' value to make it optional by adding `?` and applying the same behavior.

File: src/ng/compile.js (hunk 1 above)
Comment (petebacondarwin) [inline src/ng/compile.js]: We have to differentiate between `null` and `undefined`. I will add a comment to clarify.

**Comment by petebacondarwin** [comment]
We could make `$$slots` -> `$slots`.

**Comment by petebacondarwin** [comment]
Regarding default content for the default transclusion, perhaps we could special case `""` in the transclude map:

```js
transclude: {
  'required': 'requiredSlot',
  'optional': '?optionalSlot',
  '': '?' // this would make the default transclusion optional too...
}
```

File: src/ng/compile.js (hunk 1 above)
Comment (Narretz) [inline src/ng/compile.js]: Okay, I just noticed there's no difference in test behavior.

**Comment by Narretz** [comment]
Then we should also document `$slots`, I guess.

For default transchlusion. Okay, I forgot you can have default transclusion and slots. The proposed API for that looks a bit silly, but it's probably okay. But we can also decide on that after 1.5

**Comment by petebacondarwin** [comment]
Hmm it is a bit awkward to make the default transclusion optional. I think we should punt that feature to post-1.5.0

**Comment by Narretz** [comment]
Looks like the last commit caused some doc e2e tests to fail. :o Other than that, feel free to merge. We can tweak it after the fact.

**Pull request merged**